synapse/storage/prepare_database.py

```diff
@@ -108,9 +108,12 @@ def prepare_database(
         # so we start one before running anything. This ensures that any upgrades
         # are either applied completely, or not at all.
         #
-        # (psycopg2 automatically starts a transaction as soon as we run any statements
-        # at all, so this is redundant but harmless there.)
-        cur.execute("BEGIN TRANSACTION")
+        # psycopg2 does automatically start transactions, and while this is technically
+        # harmless to do there as well, doing so results in nested transactions. Those
+        # will generate warnings in Postgres' logs per query, and we'd rather like to
+        # avoid doing that.
+        if isinstance(database_engine, Sqlite3Engine):
```

Comment:
Given the comment above, I think I'd be a little more paranoid: maybe only skip this if we are postgres and the connection does not have autocommit set?